### ISSUE TITLE
Some merges may still stuck?

### ISSUE THREAD

**Issue by zylthinking** [ClickHouse/ClickHouse#53143]
I am analyzing the issue https://github.com/ClickHouse/ClickHouse/issues/10368, and the fixed code.
Here I have a question about searchForMissingPartOnOtherReplicas:

```c++
bool ReplicatedMergeTreePartCheckThread::searchForMissingPartOnOtherReplicas(const String & part_name) const
{
    [...]

    bool found_part_with_the_same_min_block = false;
    bool found_part_with_the_same_max_block = false;

    Strings replicas = zookeeper->getChildren(storage.zookeeper_path + "/replicas");

    [...]

    /// Check all replicas and our replica must be this last one
    for (const String & replica : replicas)
    {
        String replica_path = storage.zookeeper_path + "/replicas/" + replica;

        Strings parts = zookeeper->getChildren(replica_path + "/parts");
        Strings parts_found;
        for (const String & part_on_replica : parts)
        {
             [...]

            if (part_info.contains(part_on_replica_info))
            {
                if (part_on_replica_info.min_block == part_info.min_block)
                {
                    found_part_with_the_same_min_block = true;
                    parts_found.push_back(part_on_replica);
                }

                if (part_on_replica_info.max_block == part_info.max_block)
                {
                    found_part_with_the_same_max_block = true;

                    /// If we are looking for part like partition_X_X_level we can add part
                    /// partition_X_X_(level-1) two times, avoiding it
                    if (parts_found.empty() || parts_found.back() != part_on_replica)
                        parts_found.push_back(part_on_replica);
                }

                if (found_part_with_the_same_min_block && found_part_with_the_same_max_block)
                {
                    LOG_INFO(log, "Found parts with the same min block and with the same max block as the missing part {} on replica {}. "
                             "Hoping that it will eventually appear as a result of a merge. Parts: {}",
                             part_name, replica, fmt::join(parts_found, ", "));
                    return true;
                }
            }
        }
    }
    [...]
    return false;
}
```

There 2 variants found_part_with_the_same_min_block and found_part_with_the_same_max_block, defined outside of the loop.
Assigned and checked in the loop.

Technically, the 2 variants can be assigned in different loops.
That means, 1 replicas contains  found_part_with_the_same_min_block and the 2nd contains found_part_with_the_same_max_block.

I don't know why the 2 parts have the possibility to be merged. 

To my understanding, since `part_name` has been introduced, that means a merging has started for `part_name`, which means It is possible that GET_PARTS  for  the 2 parts has been executed,  so the 2 parts will not appear in same replica, and it is not possible for the 2 parts to be merged into `part_name`.

So the merging waiting for `part_name` be ready will still stuck.

Another question about the commence
```
/// If we are looking for part like partition_X_X_level we can add part
/// partition_X_X_(level-1) two times
```

Why it is possible?

**Comment by ljluestc** [comment]
```

bool ReplicatedMergeTreePartCheckThread::searchForMissingPartOnOtherReplicas(const String & part_name) const
{
    bool found_part_with_the_same_min_block = false;
    bool found_part_with_the_same_max_block = false;

    Strings replicas = zookeeper->getChildren(storage.zookeeper_path + "/replicas");

    // Check all replicas
    for (const String & replica : replicas)
    {
        String replica_path = storage.zookeeper_path + "/replicas/" + replica;

        Strings parts = zookeeper->getChildren(replica_path + "/parts");
        Strings parts_found;

        // Check parts on each replica
        for (const String & part_on_replica : parts)
        {
            // Simplified part_info and part_on_replica_info comparison
            if (part_info.contains(part_on_replica_info))
            {
                if (part_on_replica_info.min_block == part_info.min_block)
                {
                    found_part_with_the_same_min_block = true;
                    parts_found.push_back(part_on_replica);
                }

                if (part_on_replica_info.max_block == part_info.max_block)
                {
                    found_part_with_the_same_max_block = true;

                    // Avoid adding the same part multiple times
                    if (parts_found.empty() || parts_found.back() != part_on_replica)
                        parts_found.push_back(part_on_replica);
                }

                // If both min_block and max_block conditions are met for the same part, log and return true
                if (found_part_with_the_same_min_block && found_part_with_the_same_max_block)
                {
                    LOG_INFO(log, "Found parts with the same min block and with the same max block as the missing part {} on replica {}. "
                                 "Hoping that it will eventually appear as a result of a merge. Parts: {}",
                             part_name, replica, fmt::join(parts_found, ", "));
                    return true;
                }
            }
        }
    }

    return false;
}
`bool ReplicatedMergeTreePartCheckThread::searchForMissingPartOnOtherReplicas(const String & part_name) const
{
    bool found_part_with_the_same_min_block = false;
    bool found_part_with_the_same_max_block = false;

    Strings replicas = zookeeper->getChildren(storage.zookeeper_path + "/replicas");

    // Check all replicas
    for (const String & replica : replicas)
    {
        String replica_path = storage.zookeeper_path + "/replicas/" + replica;

        Strings parts = zookeeper->getChildren(replica_path + "/parts");
        Strings parts_found;

        // Check parts on each replica
        for (const String & part_on_replica : parts)
        {
            // Simplified part_info and part_on_replica_info comparison
            if (part_info.contains(part_on_replica_info))
            {
                if (part_on_replica_info.min_block == part_info.min_block)
                {
                    found_part_with_the_same_min_block = true;
                    parts_found.push_back(part_on_replica);
                }

                if (part_on_replica_info.max_block == part_info.max_block)
                {
                    found_part_with_the_same_max_block = true;

                    // Avoid adding the same part multiple times
                    if (parts_found.empty() || parts_found.back() != part_on_replica)
                        parts_found.push_back(part_on_replica);
                }

                // If both min_block and max_block conditions are met for the same part, log and return true
                if (found_part_with_the_same_min_block && found_part_with_the_same_max_block)
                {
                    LOG_INFO(log, "Found parts with the same min block and with the same max block as the missing part {} on replica {}. "
                                 "Hoping that it will eventually appear as a result of a merge. Parts: {}",
                             part_name, replica, fmt::join(parts_found, ", "));
                    return true;
                }
            }
        }
    }

    return false;
}
```

**Comment by intfish123** [comment]
I has same problems as yours, and some task of megers are stucked, the propress no changed very long time(10+ houres), I guess  that some parts of entries of `system.merge` need `get parts`, but `get parts` are blocked by fetched threads cause of limited `...max 8`, then it all stucked.


I resolve it with fllow commands at problem machine:
```
clickhouse-client -m --host localhost --port 9000 --user <your-username> --password <your-password>  --query "SELECT replica_path || '/queue/' || node_name FROM system.replication_queue JOIN system.replicas USING (database, table) where type in ('MERGE_PARTS', 'GET_PART');" | while read i; do clickhouse-keeper-client -h <your-keeper-host> -p 9181 -q "rm $i"; done

SYSTEM RESTART REPLICAS;
```

**Comment by intfish123** [comment]
> I has same problems as yours, and some task of megers are stucked, the propress no changed very long time(10+ houres), I guess that some parts of entries of `system.merge` need `get parts`, but `get parts` are blocked by fetched threads cause of limited `...max 8`, then it all stucked.
> 
> First: I check the network status, found that network card change Up/Down frequently, it block `get part` entity.
> 
> If network card worked normally, I will try this (Risky commands):
> 
> ```
> clickhouse-client -m --host localhost --port 9000 --user <your-username> --password <your-password>  --query "SELECT replica_path || '/queue/' || node_name FROM system.replication_queue JOIN system.replicas USING (database, table) where type in ('MERGE_PARTS', 'GET_PART');" | while read i; do clickhouse-keeper-client -h <your-keeper-host> -p 9181 -q "rm $i"; done
> 
> SYSTEM RESTART REPLICAS;
> ```

The reason is network card change Up/Down frequently, and block `get part`.
